### PR TITLE
Add diagnostic to count number of field substeps vs. track energy

### PR DESCRIPTION
**Pull request by amandalund**
@sethrj this is the field diagnostic that I updated to tally the number of substeps per track as a function of track energy (see #850). The extra state data needed (number of substeps, pre-step energy) is only allocated if the diagnostic is enabled; it shouldn't be expensive, but the changes still might be too intrusive, so we can definitely just close this without merging and keep it in the history. As an example, here's a plot of the diagnostic output from running the cms2018 regression problem:
field-diagnostic1

### CODE REVIEW

**Comment by sethrj** [comment]
beautiful

**Comment by sethrj** [comment]
That has quite a bit of interesting structure...

**Comment by amandalund** [comment]
Yeah, I'm not sure what causes it to oscillate like that...

**Comment by sethrj** [comment]
@whokion Can you scope out what would be involved for creating a similar plot for the Geant4 substep count?

File: src/celeritas/field/FieldPropagator.hh [hunk 1]
```diff
@@ -180,6 +180,7 @@ CELER_FUNCTION auto FieldPropagator<DriverT, GTV>::operator()(real_type step)
         // Advance up to (but probably less than) the remaining step length
         DriverResult substep = driver_.advance(remaining, state_);
         CELER_ASSERT(substep.step > 0 && substep.step <= remaining);
+        ++result.num_substeps;
```
Comment (sethrj) [inline src/celeritas/field/FieldPropagator.hh:183]: One other useful diagnostic would be to use the "diagnostic substepper" (or equivalent) to count the number of *integrations* as a function of energy.

**Pull request closed without merging**